File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys
myinput = sys.stdin.readline
sys.setrecursionlimit(10**9)

N = int(myinput())
graph = []
for _ in range(N):
    graph.append(list(map(int, myinput().split())))

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def dfs(x, y, num):
    if visited[x][y] == 0:
        visited[x][y] = 1
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx<0 or nx>=N or ny<0 or ny>=N:
                continue
            if visited[nx][ny]==0 and graph[nx][ny]>num:
                dfs(nx,ny,num)
        return True
    return False

num=0 # 잠기는 높이
res = []
for num in range(0,101):
    visited = [[0] * N for _ in range(N)]
    cnt=0 # 안전 영역 개수
    for i in range(N):
        for j in range(N):
            if graph[i][j]>num and dfs(i,j, num):
                cnt+=1
    res.append(cnt)
print(max(res))

File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include<iostream>
#include<queue>
#include<algorithm>
using namespace std;
int N, high = 0, low = 101, answer = 0;
int graph[100][100]; //주어진 높이 정보
bool visited[100][100];
int dx[4] = {0, 0, 1, -1};
int dy[4] = {1, -1, 0, 0};
bool dfs(int x, int y, int k) {
	if (graph[x][y] <= k || visited[x][y]) return false;
	visited[x][y] = true;
	for (int i = 0; i < 4; i++) {
		int nx = x + dx[i];
		int ny = y + dy[i];
		if (nx < 0 || nx >= N || ny < 0 || ny >= N || visited[nx][ny] || graph[nx][ny] <= k) continue;
		dfs(nx, ny, k);
	}
	return true;
}
int main() {
	ios::sync_with_stdio(0);
	cin.tie(0);
	//입력받기
	cin >> N;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			cin >> graph[i][j];
		}
		high = max(high, *max_element(&graph[i][0], &graph[i][N]));
		low = min(low, *min_element(&graph[i][0], &graph[i][N]));
	}
	//low ~ high 사이의 높이별로 dfs
	for (int k = low - 1; k < high; k++) {
		fill(&visited[0][0], &visited[99][100], false);
		//dfs돌리기
		int cnt = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (dfs(i, j, k)) {
					cnt++;
				}
			}
		}
		//cnt 맥스 비교해서 답 갱신
		answer = max(answer, cnt);
	}
	cout << answer;
	return 0;
}
